### ISSUE TITLE
FileVideoStream Problem: FATAL: exception not rethrown Aborted (core dumped)

### ISSUE THREAD

**Issue by knaffe** [jrosebr1/imutils#38]
Code:

> fvs = FileVideoStream("./1_2017-03-17_08-38-27.mp4")
> fvs.start()
> num = 0 
> while fvs.more():
> 	frame = fvs.read()
> 	if frame is None:
> 		print("Break!")
> 		break
> 	print(num)
> 	num = num +1
> print("finished!")
> fvs.stop()

First, It wouldn't print the number, that means fvs.more() doesn't work.
And it would raise a ERROR at last.
Notice, the path of video is on the way. 
How can I deal with my problem?

**Comment by jrosebr1** [comment]
What was your error and/or traceback?

**Comment by msz89** [comment]
Adrian, I'm getting a similar error:

FATAL: exception not rethrown
Aborted (core dumped)

I've opened two webcam streams, and it happens when I try break from my main while loop
```
    webcam1 = VideoStream(src=0, framerate=framerate).start()
    webcam2 = VideoStream(src=1, framerate=framerate).start()

    while True:
        total = 0
        print('[INFO] Starting bullet time...')
        label = input('[INFO] Enter label for video [q to quit]: ')
        if label in ["q","Q"]:
            cv2.destroyAllWindows()
            print('[INFO] Compressing folders into tar')
            sql_upper.make_tar_from_list('testicle', list_to_compress)
            webcam1.stop()
            webcam2.stop()
            break ##ERROR THROWN HERE
```

**Comment by jrosebr1** [comment]
Can you confirm that this is an OpenCV issue an not related to your `sql_upper` module? If not, this might be a threading issue. Try:

```
import time
time.sleep(3.0)
webcam1.stop()
webcam2.stop()
```
And see if the additional sleep allows the threads to close properly.

**Comment by msz89** [comment]
Thanks for your quick response.
I tried putting the sleep in (and commenting out my other functions) but to no avail.

Is there anything else that needs to happen to get the `VideoStream` to close off the resources its using.
A bit of google suggests that the 'Exception not rethrown' is often associated with C++ and threading - I immediately thought CV2.

From what I understand `videostream` creates a `webcamvideostream` class object.
Wrt. the `WebcamVideoStream` object we have made:
Calling the `.start()` method creates a Thread and kicks off the `.update()` class method.
Update loops continuously until `self.stopped == True` (which happens when the `.stop()` method is called on our stream.

I'm not an expert on threading, but is there a need to close of the thread `t = Thread(target=self.update, args=())` that we create in the `.start()` method? 
Could this be where the resources are held up?

I'm happy to provide more info about my use case.

Thanks again, we're getting a lot out of your book!
Matt

**Comment by jrosebr1** [comment]
> I'm not an expert on threading, but is there a need to close of the thread t = Thread(target=self.update, args=()) that we create in the .start() method?
Could this be where the resources are held up?

This is _very likely_ what is causing the problem here. The thread is likely not being cleaned up. In my use cases it seems that the thread is being properly collected, but somewhere our use cases deviate and yours is not exiting properly.

Unfortunately I'm not sure where the exact error is. I think you're on the right track though. Take a look at the comments section of these blog posts where threading is utilized:

- https://www.pyimagesearch.com/2016/01/04/unifying-picamera-and-cv2-videocapture-into-a-single-class-with-opencv/
- https://www.pyimagesearch.com/2015/12/28/increasing-raspberry-pi-fps-with-python-and-opencv/
- https://www.pyimagesearch.com/2015/12/21/increasing-webcam-fps-with-python-and-opencv/

I know readers have mentioned having threading problems before (specifically on exit) but we haven't been able to pinpoint exactly where the issue is. Threading can be a bit of a bear. I'm sorry I cannot be of more help here but if we diagnose and resolve the issue I'd be happy to update the library.

**Comment by lukasbrchl** [comment]
@knaffe I fixed this in #86

**Comment by jrosebr1** [comment]
Resolved by merging in #86.
